### PR TITLE
NullReferenceException when Partial View couldn't be found

### PR DESCRIPTION
**Pull request by dwonisch**
Issue https://github.com/NancyFx/Nancy/issues/2165

### CODE REVIEW

File: src/Nancy.Tests.Functional/Tests/PartialViewTests.cs [hunk 1]
```diff
@@ -43,5 +43,17 @@ public void When_Using_Partial_View_Then_First_Index_Of_ViewStart_Should_Equal_L
             // If the index is not the same then the string occurs twice...
             Assert.Equal(firstIndex, lastIndex);
         }
+
+        [Fact]
+        public void When_Partial_View_Could_Not_Be_Found_An_Meaningful_Exception_Should_Be_Thrown() {
+
```
Comment (khellang) [inline src/Nancy.Tests.Functional/Tests/PartialViewTests.cs]: The brace above really wants to sit here :wink:

File: src/Nancy.Tests.Functional/Tests/PartialViewTests.cs [hunk 2]
```diff
@@ -43,5 +43,17 @@ public void When_Using_Partial_View_Then_First_Index_Of_ViewStart_Should_Equal_L
...
             Assert.Equal(firstIndex, lastIndex);
         }
+
+        [Fact]
+        public void When_Partial_View_Could_Not_Be_Found_An_Meaningful_Exception_Should_Be_Thrown() {
+
+            Assert.Throws<ViewNotFoundException>(() =>
+            {
+                var response = browser.Get(@"/razor-partialnotfound", with =>
+                { with.HttpRequest(); });
```
Comment (khellang) [inline src/Nancy.Tests.Functional/Tests/PartialViewTests.cs]: Is HttpRequest necessary? I think that's the default.

File: src/Nancy.ViewEngines.Razor/HtmlHelpers.cs [hunk 3]
```diff
@@ -2,6 +2,7 @@
 {
     using System;
     using System.IO;
+    using System.Linq;
```
Comment (khellang) [inline src/Nancy.ViewEngines.Razor/HtmlHelpers.cs]: I don't think this is needed?

File: src/Nancy.ViewEngines.Razor/HtmlHelpers.cs [hunk 4]
```diff
@@ -16,8 +17,7 @@ public class HtmlHelpers<TModel> : HtmlHelpers
         /// <param name="engine">The razor view engine instance that the helpers are being used by.</param>
         /// <param name="renderContext">The <see cref="IRenderContext"/> that the helper are being used by.</param>
         /// <param name="model">The model that is used by the page where the helpers are invoked.</param>
-        public HtmlHelpers(RazorViewEngine engine, IRenderContext renderContext, TModel model) : base(engine, renderContext)
-        {
+        public HtmlHelpers(RazorViewEngine engine, IRenderContext renderContext, TModel model) : base(engine, renderContext) {
```
Comment (khellang) [inline src/Nancy.ViewEngines.Razor/HtmlHelpers.cs]: Revert this change, thanks :smile:

File: src/Nancy.ViewEngines.Razor/HtmlHelpers.cs [hunk 5]
```diff
@@ -38,6 +38,7 @@ public abstract class HtmlHelpers
         /// </summary>
         /// <param name="engine">The razor view engine instance that the helpers are being used by.</param>
         /// <param name="renderContext">The <see cref="IRenderContext"/> that the helper are being used by.</param>
+        /// <param name="factory"></param>
```
Comment (khellang) [inline src/Nancy.ViewEngines.Razor/HtmlHelpers.cs]: I don't :eyes: this parameter anywhere...

File: src/Nancy.ViewEngines.Razor/HtmlHelpers.cs [hunk 6]
```diff
@@ -56,6 +57,8 @@ protected HtmlHelpers(RazorViewEngine engine, IRenderContext renderContext)
         /// <value>An <see cref="IRenderContext"/> instance.</value>
         public IRenderContext RenderContext { get; set; }
 
+        private IViewFactory ViewFactory { get; set; }
```
Comment (khellang) [inline src/Nancy.ViewEngines.Razor/HtmlHelpers.cs]: This seems... unused?

File: src/Nancy/ViewEngines/ViewNotFoundException.cs [hunk 7]
```diff
@@ -39,6 +39,27 @@ public ViewNotFoundException(string viewName, string[] availableViewEngineExtens
...
         }
 
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ViewNotFoundException"/>.
+        /// </summary>
+        /// <param name="viewName">The name of the view that was being located.</param>
+        /// <param name="availableViewEngineExtensions">List of available view extensions that can be rendered by the available view engines.</param>
+        /// <param name="inspectedLocations">The locations that were inspected for the view.</param>
+        /// <param name="rootPathProvider">An <see cref="IRootPathProvider"/> instance.</param>
```
Comment (khellang) [inline src/Nancy/ViewEngines/ViewNotFoundException.cs]: `inspectedLocations` and `rootPathProvider` don't need a place here.

File: src/Nancy/ViewEngines/ViewNotFoundException.cs [hunk 8]
```diff
@@ -39,6 +39,27 @@ public ViewNotFoundException(string viewName, string[] availableViewEngineExtens
                     Environment.NewLine);
         }
 
+
```
Comment (khellang) [inline src/Nancy/ViewEngines/ViewNotFoundException.cs]: :fire: this extra blank line

File: src/Nancy/ViewEngines/ViewNotFoundException.cs [hunk 9]
```diff
@@ -39,6 +39,27 @@ public ViewNotFoundException(string viewName, string[] availableViewEngineExtens
...
+            this.AvailableViewEngineExtensions = availableViewEngineExtensions;
+
+            this.message = String.Format(
+                    "Unable to locate view '{0}'{2}Currently available view engine extensions: {1}{2}",
+                    this.ViewName,
+                    string.Join(",", this.AvailableViewEngineExtensions),
+                    Environment.NewLine);
+        }
+
+
```
Comment (khellang) [inline src/Nancy/ViewEngines/ViewNotFoundException.cs]: :fire: this extra blank line

File: src/Nancy.ViewEngines.Razor/HtmlHelpers.cs [hunk 10]
```diff
@@ -76,6 +79,9 @@ public IHtmlString Partial(string viewName, dynamic modelForPartial)
         {
             var view = this.RenderContext.LocateView(viewName, modelForPartial);
 
+            if (view == null)
+                throw new ViewNotFoundException(viewName, Engine.Extensions.ToArray());
```
Comment (khellang) [inline src/Nancy.ViewEngines.Razor/HtmlHelpers.cs]: Can you add braces around this?

**Comment by khellang** [comment]
Thanks for this @dwonisch! :grin::+1:

I left some comments for you. We'd :heart: if you could tweak those things and squash the commits.

**Comment by dwonisch** [comment]
Updated the changes. Would be nice to have a VS Settings file or something like that, to automatically change this settings :)

**Comment by khellang** [comment]
> Would be nice to have a VS Settings file or something like that, to automatically change this settings :)

We have both a .DotSetting (for ReSharper) and a .editorconfig file. What kind of VS settings file did you have in mind?

**Comment by khellang** [comment]
This closes #2165

**Comment by thecodejunkie** [comment]
@ericrrichards We'll be sneaking this into the (very) pending `1.4.3` release instead :D

**Comment by khellang** [comment]
Version 1.4.3 was just pushed to [NuGet](https://www.nuget.org/packages/Nancy/1.4.3).
